First attempt to refactor button stylesheet

Stanton opened the pull request:
Target branch = `purple`

This is a first-pass refactor of the button component which aims to simplify the styles required to generate Pulsar's buttons and reduce overall specificity.

It is expected that there will be additional work required before this PR is acceptable.

Notable changes:
* Bottom outset border has been removed
* All reliance on `!important` has been removed
* Font size increased to 18px
* Button target height increased to 38px to match planned input sizes
* Outline button borders are now 1px instead of 2px

Other notes:
* Table link styles have been stripped out in this change, fully expect additional work to happen for table links as part of Purple
* I plan on removing the `info` blue style due to lack of adoption, but as that doesn't just affect buttons I'll do this as a separate PR

### Specificity change
<img width="957" alt="button-after" src="https://user-images.githubusercontent.com/18653/107369712-038e3c00-6ada-11eb-8e52-88062f07887a.png">

### Screenshots
![Screenshot 2021-02-09 at 12 57 37](https://user-images.githubusercontent.com/18653/107366835-7dbcc180-6ad6-11eb-96ba-7b5dca972d5f.png)
![Screenshot 2021-02-09 at 12 57 43](https://user-images.githubusercontent.com/18653/107366843-81504880-6ad6-11eb-8f84-657c6c54c8bb.png)

jamesjacobs commented on stylesheets/_component.buttons.scss:
```diff
@@ -1,385 +1,110 @@
-$btn-padding-top:           10px;
-$btn-padding-bottom:        7px;
-$btn-padding-x:             14px;
+// Variables
```
Move to `_config.variables.scss`?

Stanton replied:
Will be moving when things stabilise

jamesjacobs commented on stylesheets/_component.buttons.scss:
```diff
@@ -1,385 +1,110 @@
 
+// State variations
 @each $state, $state-color, $state-color-alt in $state-colors {
     .btn--#{'' + $state} {
```
Wrap in `@if $state != 'base' {}` to avoid unused `--base` styles being outputted in compiled CSS.

jamesjacobs commented on stylesheets/_component.buttons.scss:
```diff
@@ -1,385 +1,110 @@
-    &.btn--small {
-        padding: ($btn-small-padding-top - 2) ($btn-small-padding-x - 2) ($btn-small-padding-bottom - 2) !important;
+    background: none;
+    border: 1px solid color(base);
```
See notes on this. Feels like 1px is too thin.

Stanton replied:
I'm going to run with it for now, most notable design systems use 1px for outline buttons and I'm hoping this will reduce our reliance on naked buttons a little by being subtle enough to not draw too much focus

Stanton replied:
Leaving for now

jamesjacobs commented on stylesheets/_component.buttons.scss:
```diff
@@ -1,385 +1,110 @@
-    -webkit-appearance: none; /* 1 */
-    background: transparent;
+    @include button-variant('base');
     border: 0;
```
See notes on button borders.

jamesjacobs commented on stylesheets/_component.buttons.scss:
```diff
@@ -1,385 +1,110 @@
     }
 
     @each $state, $state-color, $state-color-alt in $state-colors {
         &.btn--#{'' + $state} {
```
Wrap in `@if $state != 'base'` to avoid outputting unused `--base` styles in compiled CSS.

jamesjacobs commented on stylesheets/_component.tables.scss:
```diff
@@ -17,33 +17,33 @@
-    [href].btn {
-        border-bottom: 0;
-    }
+    // [href],
```
Assume we're looking to replace these styles with a class?

Stanton replied:
Maaaybe, it's one to discuss for sure

jamesjacobs reviewed (commented):
### Notes & thoughts so far

**Standard buttons**
- Default grey button without a border or outline may look a little lost on a grey piano or reports UI. While not a WCAG requirement, it's recommended that component boundaries  are delineated  (see boundaries https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html)

**Outline buttons**
- Slight visual shift on focus
- Border size reduction? Harder to see , especially on coloured backgrounds such as panels.

**Compiled CSS**
- `btn--base` styles are still present, contents of each loop needs wrapping in `@if $state != 'base'`
- All button styles are duplicated due to `@import 'component.buttons'` in `_component.remove-button.scss` - removing that @import doesn't seem to cause any issues in lexicon - needs further testing.
- Duplication of styles and `--base` styles in compiled CSS also present in label CSS. Might be out of scope for this PR but potentially worth fixing at the same time.

**Questions:**
- PR mentions increasing height to 38px - should we consider bumping this to 44px to meet AAA on 2.5.5 Target size (https://www.w3.org/WAI/WCAG21/Understanding/target-size)

Stanton commented:
> * Default grey button without a border or outline may look a little lost on a grey piano or reports UI. While not a WCAG requirement, it's recommended that component boundaries  are delineated  (see boundaries https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html)
Changed default outline button to be same as inverse for the time being

> * Slight visual shift on focus
Will tweak

> * Border size reduction? Harder to see , especially on coloured backgrounds such as panels.
Will cross those bridges as they come up, might be that most of the time we shouldn't/wouldn't want to use an outline button on a non-white background, panels/flash might be exceptions though

> * `btn--base` styles are still present, contents of each loop needs wrapping in `@if $state != 'base'`
Good shout, added

> * All button styles are duplicated due to `@import 'component.buttons'` in `_component.remove-button.scss` - removing that @import doesn't seem to cause any issues in lexicon - needs further testing.
Will remove that @import as part of this PR, will watch out for any oddities, looks like it was a brute force specificity override in the past

> * Duplication of styles and `--base` styles in compiled CSS also present in label CSS. Might be out of scope for this PR but potentially worth fixing at the same time.
Not touching labels here

> * PR mentions increasing height to 38px - should we consider bumping this to 44px to meet AAA on 2.5.5 Target size (https://www.w3.org/WAI/WCAG21/Understanding/target-size)
44 is massive, legit massive, I can't bring myself to do it, we're not intending to meet AAA for anything else so I can't really justify the amount of work it would take to bring all other components up to this minimum size

Stanton commented:
Replaced by #1380
